Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,365 @@
+    /// creating connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(request: HTTPClient.Request) {
+            switch request.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
+                fatalError("HTTPClient.Request scheme should already be a valid one")
+            }
+            self.port = request.port
+            self.host = request.host
+        }
+
+        var scheme: Scheme
+        var host: String
+        var port: Int
+
+        enum Scheme: Hashable {
+            case http
+            case https
+        }
+    }
+
+    class Connection {
+        init(key: Key, channel: Channel, parentPool: ConnectionPool) {
+            self.key = key
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        func release() {
+            self.parentPool.release(self)
+        }
+
+        let parentPool: ConnectionPool
+        fileprivate let key: Key
+        let channel: Channel
+        var isLeased: Bool = false
+    }
+
+    /// This enum abstracts the underlying kind of provider
+    enum ConnectionProvider {
+        case http1(HTTP1ConnectionProvider)
+        indirect case future(EventLoopFuture<ConnectionProvider>)
+
+        func getConnection(on eventLoop: EventLoop) -> EventLoopFuture<Connection> {
+            switch self {
+            case .http1(let provider):
+                return provider.getConnection(preference: .delegate(on: eventLoop))
+            case .future(let futureProvider):
+                return futureProvider.flatMap { provider in
+                    provider.getConnection(on: eventLoop)
+                }
+            }
+        }
+
+        func release(connection: Connection) {
+            switch self {
+            case .http1(let provider):
+                return provider.release(connection: connection)
+            case .future(let futureProvider):
+                futureProvider.whenSuccess { provider in
+                    provider.release(connection: connection)
+                }
+            }
+        }
+
+        func closeAllConnections() -> EventLoopFuture<Void> {
+            switch self {
+            case .http1(let provider):
+                return provider.closeAllConnections()
+            case .future(let futureProvider):
+                return futureProvider.flatMap { provider in
+                    provider.closeAllConnections()
+                }
+            }
+        }
+    }
+
+    class HTTP1ConnectionProvider {
+        let loopGroup: EventLoopGroup
+        let configuration: HTTPClient.Configuration
+        let key: ConnectionPool.Key
+        var isClosed: Atomic<Bool>
+        /// The curently opened, unleased connections in this connection provider
+        var availableConnections: CircularBuffer<Connection> = .init(initialCapacity: 8) {
```

Comment:
this needs documentation on what the intended synchronisation strategy here is. As of right now, this is accessed from random threads and without a lock.

Reply:
Thanks, I will do this, I'm going to try placing all of the data structures that need synchronisation in a separate data structure as we discussed.